### PR TITLE
Add timeouts to SimpleConsumerWrapper

### PR DESCRIPTION
**Pull request by jfim**
Add timeouts to SimpleConsumerWrapper instead of ignoring the
timeoutMillis parameter. Add documentation for some methods which did
not have any.

### CODE REVIEW

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 1]
```diff
@@ -232,8 +233,13 @@ protected void setupInitialSegments(String realtimeTableName, ZNRecord partition
 
         metadata.setCreationTime(now);
 
-        final long startOffset = kafkaConsumer.fetchPartitionOffset(initialOffset,
-            KAFKA_PARTITION_OFFSET_FETCH_TIMEOUT_MILLIS);
+        final long startOffset;
+        try {
+          startOffset = kafkaConsumer.fetchPartitionOffset(initialOffset, KAFKA_PARTITION_OFFSET_FETCH_TIMEOUT_MILLIS);
+        } catch (TimeoutException e) {
+          LOGGER.warn("Timed out when fetching partition offsets");
```
Comment (antumbde) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: log topic name

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 2]
```diff
@@ -197,7 +198,14 @@ protected void consumeLoop() {
     while(!_receivedStop && !endCriteriaReached()) {
       // Consume for the next _kafkaReadTime ms, or we get to final offset, whichever happens earlier,
       // Update _currentOffset upon return from this method
-      Iterable<MessageAndOffset> messagesAndOffsets = _consumerWrapper.fetchMessages(_currentOffset, _endOffset, KAFKA_MAX_FETCH_TIME_MILLIS);
+      Iterable<MessageAndOffset> messagesAndOffsets = null;
+      try {
+        messagesAndOffsets = _consumerWrapper.fetchMessages(_currentOffset, _endOffset, KAFKA_MAX_FETCH_TIME_MILLIS);
+      } catch (TimeoutException e) {
+        LOGGER.warn("Timed out when fetching messages from Kafka, retrying");
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java]: topic name

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 3]
```diff
@@ -333,21 +334,25 @@ private void setCurrentState(State newState) {
...
       // Try to get into a state where we're connected to Kafka
-      // TODO This needs a time limit, and perhaps a back-off before we query kafka again.
-      while (!_currentState.isConnectedToKafkaBroker()) {
+      while (!_currentState.isConnectedToKafkaBroker() && System.currentTimeMillis() < endTime) {
         _currentState.process();
       }
 
-      // Send the metadata request to Kafka
+      if (endTime <= System.currentTimeMillis() && !_currentState.isConnectedToKafkaBroker()) {
+        throw new TimeoutException();
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: add info about topic or table name in timeoutexception

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java [hunk 4]
```diff
@@ -385,14 +390,35 @@ public synchronized int getPartitionCount(String topic, long maxWaitTimeMs) {
...
+  public synchronized Iterable<MessageAndOffset> fetchMessages(long startOffset, long endOffset, int timeoutMillis)
+      throws java.util.concurrent.TimeoutException {
     // Ensure that we're connected to the leader
-    // TODO Add a timeout/error handling
-    while(_currentState.getStateValue() != ConsumerState.CONNECTED_TO_PARTITION_LEADER) {
+    // TODO Improve error handling
+
+    final long endTime = System.currentTimeMillis() + timeoutMillis;
+
+    while(_currentState.getStateValue() != ConsumerState.CONNECTED_TO_PARTITION_LEADER &&
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: seems repeated pattern that can be moved to a method taking Function returning bool as param

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java (hunk 1 above)
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: Better yet,  log segment name. it has the partition ID and table name,

File: pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java (hunk 2 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java]: same

Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java]: actually, you can use segmentlogger here.

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java (hunk 3 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/kafka/SimpleConsumerWrapper.java]: and partition number. Perhaps it is best to instantiate a separate logger like we have done in LLSegmentDataManager that has topic + partition in it?

**Comment by mcvsubbu** [comment]
Don't you need the timeout settings in ConnectingToBootstrapNode.process() and FetchingLeaderInformation.process(), and ConnectingToPartitionLeader.process() method as well?

**Comment by jfim** [comment]
Not really, those methods don't block for an extended period of time (other than socket connection establishment and the 100ms sleep period at the end if we need to introduce a delay between retries).

Because the loop in each function manages the overall timeout of operations, the worst that can happen is that we slightly overrun the requested timeout, but none of this code is latency sensitive to the point where an extra 99ms will break it.

**Comment by mcvsubbu** [comment]
lgtm. Looks like I cannot approve this because it was submitted before we had the approval stuff?

**Pull request merged**